pkg/server/registry/apigroups/acorn/jobs/restart.go

```diff
@@ -0,0 +1,102 @@
+	apiv1 "github.com/acorn-io/runtime/pkg/apis/api.acorn.io/v1"
+	kclient "github.com/acorn-io/runtime/pkg/k8sclient"
+	batchv1 "k8s.io/api/batch/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func NewRestart(c client.WithWatch) rest.Storage {
+	return stores.NewBuilder(c.Scheme(), &apiv1.JobRestart{}).
+		WithCreate(&restartStrategy{client: c}).
+		Build()
+}
+
+type restartStrategy struct {
+	client client.WithWatch
+}
+
+func (s *restartStrategy) New() types.Object {
+	return &apiv1.JobRestart{}
+}
+
+func (s *restartStrategy) Create(ctx context.Context, obj types.Object) (types.Object, error) {
+	ri, _ := request.RequestInfoFrom(ctx)
+
+	if ri.Name == "" || ri.Namespace == "" {
+		return obj, nil
+	}
+	key := kclient.ObjectKey{Name: ri.Name, Namespace: ri.Namespace}
+
+	cronToRestart := &batchv1.CronJob{}
+	err := s.client.Get(ctx, key, cronToRestart)
+	if err == nil {
+		return obj, s.restartCronJob(ctx, cronToRestart)
+	}
+
+	if !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+
+	jobToRestart := &batchv1.Job{}
+	err = s.client.Get(ctx, key, jobToRestart)
+	if err != nil {
+		return nil, err
+	}
+
+	return obj, s.client.Delete(ctx, jobToRestart)
+}
+
+func (s *restartStrategy) restartCronJob(ctx context.Context, cron *batchv1.CronJob) error {
+	// Find all active jobs and delete them to make way for new ones.
+	for _, jobRef := range cron.Status.Active {
+		job := &batchv1.Job{}
+		err := s.client.Get(ctx, kclient.ObjectKey{Name: jobRef.Name, Namespace: jobRef.Namespace}, job)
+		if err != nil {
+			return err
+		}
+
+		if err := s.client.Delete(ctx, job); err != nil {
+			return err
+		}
+	}
+
+	// Create a new job from the cron's template to "restart" it. This is necessary because
+	// the cronjob controller will not create a new job until the next scheduled run. By creating
+	// it here we can ensure that the job is restarted immediately.
+	template := cron.Spec.JobTemplate
+	spec := template.Spec
+	spec.TTLSecondsAfterFinished = new(int32) // Want to delete the job immediately after it finishes.
+	newJob := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: cron.Name + "-",
+			Namespace:    cron.Namespace,
+			Labels:       template.Labels,
+			Annotations:  template.Annotations,
+			OwnerReferences: []metav1.OwnerReference{{
+				APIVersion: batchv1.SchemeGroupVersion.String(),
+				Kind:       "CronJob",
+				Name:       cron.Name,
+				UID:        cron.UID,
+			}},
+		},
+		Spec: spec,
+	}
+	if err := s.client.Create(ctx, newJob); err != nil {
+		return err
+	}
+
+	// Update the cronjob status to reflect the new job being created for it.
+	s.client.Get(ctx, kclient.ObjectKey{Name: cron.Name, Namespace: cron.Namespace}, cron)
+	cron.Status.LastScheduleTime = &metav1.Time{Time: newJob.CreationTimestamp.Time}
```

Comment:
There is a gap where the CronJob is not really aware that the Job we create above is owned by it. As a result things like `Active` are not populate or (if we fill it) are not cleaned-up as Jobs go away. I will be thinking through this but would love any advice.